Add a `nix` flake

igor-ramazanov opened the pull request:
https://github.com/tale/headplane/issues/95

Adds a `nix` flake with:
* Development environment (typescript LSP, NodeJS, pnpm, go).
* Packages (`headplane` and ~`hp_agent`~ `headplane-agent`).
* NixOS `services.headplane.*` module including `services.headplane.agent.*`.

Feel free to push or borrow anything to / from this branch.

~Can't build a package at the moment, marked each problem with a `TODO` comment.~

- [x] Build a `nix` package.
- [x] Build an agent package.
- [x] Synchronize the `npm` and `nix` package version.
- [x] Provide a NixOS module.
- [x] Include agent in the NixOS module.
- [ ] ~Harden `systemd` NixOS service~ scoped out.
- [x] Documentation.
- [ ] ~NixOS module tests~ scoped out.
- [ ] ~Make non-flake friendly~ scoped out.
- [x] Build Nix packages in CI.

igor-ramazanov commented:
Would be nice to drop the requirement of having the `.git` directory as it complicates the build.

tale commented:
> Would be nice to drop the requirement of having the `.git` directory as it complicates the build.

I'm thinking of switching the semver to be stored in `package.json`. The Vite build config would just directly read it from there. Is this a better system?

igor-ramazanov commented:
@tale

> Is this a better system?

From a `nix` perspective - yes, absolutely. Then the version could be read from the `package.json`.

Context:
`nix` requires [`version`](https://github.com/tale/headplane/pull/132/commits/2291284e43a1d4d47808e99d7f1438dceafa66b3) being passed explicitly on building.

Also, the reason why it's difficult to rely on `.git/` in `nix` is because `nix` heavily encourages reproducibility - if someone else would clone the repo, build it with `nix`, then, hopefully, the output would be bit-by-bit identical if built on another machine on the same `git` hash.

And `.git/`s content can be different between developers or `git` versions.

igor-ramazanov commented:
@tale Out of curiosity - is it possible to not to require having `node_modules` after building? Was there a problem why it's required?

tale commented:
I made some progress on bundling all the dependencies, but honestly I stopped because it isn't my priority at the moment. I did want to bundle everything to eliminate `node_modules/` which could then allow me to package it all as an executable using a node executable compiler.

I don't necessarily think there's anything stopping me other than dealing with bundling headaches which ultimately is not the highest priority at the moment.

igor-ramazanov commented:
@tale So far, it works for me, using this PR already and removed custom overlays and NixOS module from my `nix` config.

Haven't verified the `agent`, as there wasn't `agent` button in the `Settings` tab, but I defined all required options like it's written in the docs.

Would be nice if someone else could double-check everything is alright.

Didn't harden the systemd security, but maybe something for the future

tale commented on docs/Nix.md:
```diff
@@ -0,0 +1,98 @@
+# Nix
```
Just a little thing, we should definitely link to this in `README.md`, please change it as you see fit, once this PR merges I'll be cutting a new release anyways.

igor-ramazanov replied:
Done

tale commented on nix/package.nix:
```diff
@@ -0,0 +1,50 @@
+    mkdir -p $out/{bin,share/headplane}
+    cp -r {build,node_modules} $out/share/headplane/
+    sed -i 's;/build/source/node_modules/react-router/dist/development/index.mjs;react-router;' $out/share/headplane/build/headplane/server.js
+    sed -i 's;define_process_env_default.PORT;process.env.PORT;' $out/share/headplane/build/headplane/server.js
```
Why is this (and probably also the line above it needed)? From this line, my understanding is that `process.env.PORT` doesn't work as an option anyways, since the port settings are now determined by the `HEADPLANE_SERVER__PORT` variable option.

igor-ramazanov replied:
It was a leftover from older versions https://github.com/tale/headplane/issues/95#issue-2807487849.

Checked and not needed indeed.

The one upper with `react-router` also doesn't do anything useful, apparently, removed too.

tale commented on version:
```diff
@@ -0,0 +1 @@
+0.5.3-SNAPSHOT
```
You can drop this file (and also please drop the changes you made in `vite.config.ts`. I added the semver in `package.json`, feel free to use it exactly as you see fit. Your best bet is probably reading it in from `jq` if this is possible.

igor-ramazanov replied:
Fixed

tale commented on nix/module.nix:
```diff
@@ -0,0 +1,115 @@
+        after = ["headplane.service"];
+        requires = ["headplane.service"];
+
+        environment = {
```
These variables have changed to reflect the new variable format I'm using for Headplane. There is also a new debug variable, could you please add these changes!

igor-ramazanov replied:
Updated

igor-ramazanov replied:
I've removed hardcoded env var names, now `services.headplane.agent.settings` will accept `nix` attribute set which then will be converted to environment variables.

tale reviewed (changes requested):
Overall this is very good, just a few small things and we should be good to go. Once again thank you for the contribution, this is a pretty big one as I know quite a few people who've been asking for this.

igor-ramazanov commented:
@tale I've addressed all comments.

In the future I may also improve `systemd` security.

* https://github.com/synacktiv/shh
* https://github.com/alegrey91/systemd-service-hardening
* https://documentation.suse.com/smart/security/html/systemd-securing/index.html

tale commented:
All good, last thing is actually a question about CI. Is it possible to use it to test the flake. I don't have nix locally so I can't really manage this repeatedly across updates.

igor-ramazanov commented:
@tale Yeah, I've been using some of Nix CI Github Actions in the past.

If by _testing_ you mean _it builds packages without errors_, but not _packages are working correctly_, then it's fairly easy to achieve and I can push it.

Ideally, the NixOS module should be tested too by spinning up a virtual machine with Headscale and Headplane, and [verifying Headplane works correctly](https://nixos.org/manual/nixos/stable/#ssec-machine-objects), but I'd scope it out from this PR.

tale commented:
Yeah this is good, I'll merge this in once I cut the next release. Thank you for the PR!
